execution_engine/src/storage/transaction_source/db.rs

```diff
@@ -0,0 +1,309 @@
+    fn migration_column_family(&self) -> Result<Arc<rocksdb::BoundColumnFamily>, error::Error> {
+        let cf = match self.db.cf_handle(MIGRATION_COLUMN_FAMILY) {
+            Some(cf) => cf,
+            None => {
+                info!(
+                    "creating column family {} for migration",
+                    MIGRATION_COLUMN_FAMILY
+                );
+                self.db
+                    .create_cf(MIGRATION_COLUMN_FAMILY, &rocksdb_defaults())?;
+                match self.db.cf_handle(MIGRATION_COLUMN_FAMILY) {
+                    Some(cf) => cf,
+                    None => {
+                        return Err(error::Error::RocksDb(format!(
+                            "unable to open column family {}",
+                            MIGRATION_COLUMN_FAMILY
+                        )))
+                    }
+                }
+            }
+        };
+        Ok(cf)
+    }
+
+    /// Check if a state root has been marked as migrated from lmdb to rocksdb.
+    pub fn is_state_root_migrated(&self, state_root: &[u8]) -> Result<bool, error::Error> {
+        Ok(self
+            .db
+            .get_cf(&self.migration_column_family()?, state_root)?
+            .is_some())
+    }
+}
+
+impl Transaction for RocksDb {
+    type Error = rocksdb::Error;
+
+    type Handle = RocksDb;
+
+    fn commit(self) -> Result<(), Self::Error> {
+        // NO OP as rockdb doesn't use transactions.
+        Ok(())
+    }
+}
+
+impl Readable for RocksDb {
+    fn read(&self, _handle: Self::Handle, key: &[u8]) -> Result<Option<Bytes>, Self::Error> {
+        let maybe_value: Option<Vec<u8>> = if self.in_mem {
+            self.mem_map.lock().unwrap().get(key).cloned()
+        } else {
+            self.db.get(key)?
+        };
+        Ok(maybe_value.map(|some| {
+            let value = some.as_ref();
+            Bytes::from(value)
+        }))
+    }
+}
+
+impl Writable for RocksDb {
+    fn write(
+        &mut self,
+        _handle: Self::Handle,
+        key: &[u8],
+        value: &[u8],
+    ) -> Result<(), Self::Error> {
+        if self.in_mem {
+            self.mem_map
+                .lock()
+                .unwrap()
+                .insert(key.to_vec(), value.to_vec());
+        } else {
+            let _result = self.db.put(key, value)?;
+        }
+        Ok(())
+    }
+}
+
+/// Environment for rocksdb.
+#[derive(Clone)]
+pub struct RocksDbStore {
+    pub(crate) rocksdb: RocksDb,
+    pub(crate) path: PathBuf,
+}
+
+impl RocksDbStore {
+    /// Create a new environment for alternative db.
+    pub fn new(
+        path: impl AsRef<Path>,
+        rocksdb_opts: rocksdb::Options,
+    ) -> Result<RocksDbStore, rocksdb::Error> {
+        let db = Arc::new(rocksdb::DBWithThreadMode::<MultiThreaded>::open(
+            &rocksdb_opts,
+            path.as_ref(),
+        )?);
+
+        let rocksdb = RocksDb {
+            db,
+            // This is just for testing.
```

Comment:
Because I'd like to delete InMemoryGlobalState in a future PR, along with quite a bit of generic code that attempts to abstract over *GlobalState. EngineState need not be so generic, nor WasmTestBuilder. I think this change would allow a great deal of simplification of both production and test code in the future.

Reply:
Removed now, talking with @xcthulhu , I'll come back to this in a future PR.